packages/sync/src/utils/withDeferredSyncSafeguards.ts

```diff
@@ -0,0 +1,41 @@
+import { BaseDatabase } from '@tupaia/database';
+
+import { switchTombstoneTriggers } from './switchTombstoneTriggers';
+
+export const withDeferredSyncSafeguards = async <T>(
+  database: BaseDatabase,
+  operation: () => Promise<T>,
+): Promise<T> => {
+  /**
+   * Defer foreign key constraint assertions until the end of the transaction.
+   *
+   * This prevents constraint violations during data synchronization when dealing with
+   * self-referencing foreign keys (e.g., entity.parent_id, permission_group.parent_id).
+   *
+   * Without deferral, these constraints would be checked immediately upon insertion,
+   * causing failures when parent records haven't been inserted yet. The hierarchical
+   * nature of this data makes it complicated to guarantee correct insertion order.
+   *
+   * Note: Only constraints explicitly altered to be DEFERRABLE will be affected.
+   * Standard foreign key constraints continue to be enforced immediately.
+   */
+  await database.executeSql(`
+    SET CONSTRAINTS ALL DEFERRED;
+  `);
+
+  // Disable tombstone triggers to avoid syncing deletion
+  // also put the records into tombstone table
+  await switchTombstoneTriggers(database, false);
```

Comment:
During refinement, we did discuss about 2 solutions to avoid putting the sync deleted records into tombstone table that create a loop of pushing deleted records
- Temporarily disable the tombstone trigger when persisting synced data
- We can also sync the tombstone  table between client and central, when the tombstone record already exists, don't reinsert it.

After taking a stab, the first solution appears to be cleaner to me because:
- Syncing tombstone data would also require transferring tombstone data into the sync_lookup table, which to me can be a lot of load
- We'll always have to persist tombstone first when syncing, which makes the code a bit awkward imo

With disabling the triggers, it makes the code a lot simpler